test/gossiper: add reproducible test for race condition during node d…

emdroid opened the pull request:
Populate the local state during gossiper initialization in start_gossiping, preventing an empty state from being added to _endpoint_state_map and returned in get_endpoint_states responses, that was causing an 'empty host id issue' on the other nodes during nodes restart.

Check for a race condition in do_apply_state_locally In do_apply_state_locally, a race condition can occur if a task is suspended at a preemption point while the node entry is not locked.
During this time, the host may be removed from _endpoint_state_map. When the task resumes, this can lead to inserting an entry with an empty host ID into the map, causing various errors, including a node crash.

This change adds a check after locking the map entry: if a gossip ACK update does not contain a host ID, we verify that an entry with that host ID still exists in the gossiper’s _endpoint_state_map.

Fixes https://github.com/scylladb/scylladb/issues/25831
Fixes https://github.com/scylladb/scylladb/issues/25803
Fixes https://github.com/scylladb/scylladb/issues/25702
Fixes https://github.com/scylladb/scylladb/issues/25621

Ref https://github.com/scylladb/scylla-enterprise/issues/5613

Backport: The issue affects all current releases(2025.x), therefore this PR needs to be backported to all 2025.1-2025.3.

emdroid commented:
This PR is re-applying the previously reverted fixes in https://github.com/scylladb/scylladb/issues/25803 and the additional fixes from https://github.com/scylladb/scylladb/pull/25844

emdroid commented:
Also run the following tests 1000 times locally (dev) to make sure the re-applied commits are fixed and won't cause CI failures:
* cluster/test_raft_cluster_features::test_joining_old_node_fails
* cluster/test_raft_recovery_basic::test_raft_recovery_basic
* cluster/test_cluster_features::test_downgrade_after_successful_upgrade_fails

Those were the tests that were failing the CI in https://github.com/scylladb/scylladb/issues/25803.

mykaul commented:
@gleb-cloudius - please review.
@scylladb/scylla-maint - please review.

calebxyz commented:
please merge this fix before merging https://github.com/scylladb/scylladb/pull/25849

gleb-cloudius commented:
I am confused about this PR. It contains reverted patches and https://github.com/scylladb/scylladb/pull/25844. May be https://github.com/scylladb/scylladb/pull/25844 should simple include revers on reverts.

calebxyz commented on gms/gossiper.cc:
```diff
@@ -2062,6 +2085,15 @@ void gossiper::examine_gossiper(utils::chunked_vector<gossip_digest>& g_digest_l
     }
 }
 
+endpoint_state gossiper::make_initial_endpoint_state(const gms::generation_type generation_nbr, const application_state_map& preload_local_states) {
```
```suggestion
endpoint_state gossiper::make_initial_endpoint_state(gms::generation_type generation_nbr, const application_state_map& preload_local_states) {
```

emdroid commented:
New version:
* changed the constness mismatch between the function declaration and definition
* const corrected the range-for loop in `make_initial_endpoint_state()`
* added the reference of the reverts and re-apply to the commit messages of the re-applied commits

emdroid commented:
> > I am confused about this PR. It contains reverted patches and #25844. May be #25844 should simple include revers on reverts.
> 
> seems they merged everything to one big PR, but its separated nicly by commits, in my opinion because this is a high priority task we should push it as is.

Yes. I could also revert the reverts instead, but I at least added the reference to the reverts and the original commits to make it clear that it is the case of re-applying these reverted commits.

gleb-cloudius commented:
OK. I will approve, but then we need to close https://github.com/scylladb/scylladb/pull/25844. We did a big mess here :(

emdroid commented:
The CI failures seem to be because of https://github.com/scylladb/scylla-dtest/pull/6167 (now reverted in https://github.com/scylladb/scylla-dtest/pull/6196).

calebxyz commented:
> The CI failures seem to be because of [scylladb/scylla-dtest#6167](https://github.com/scylladb/scylla-dtest/pull/6167) (now reverted in [scylladb/scylla-dtest#6196](https://github.com/scylladb/scylla-dtest/pull/6196)).

How does LWT test with Tablets makes this test fail ?

avikivity commented on gms/gossiper.cc:
```diff
@@ -2093,7 +2093,7 @@ future<> gossiper::start_gossiping(gms::generation_type generation_nbr, applicat
+    endpoint_state& local_state = my_endpoint_state();
     local_state.set_heart_beat_state_and_update_timestamp(heart_beat_state(generation_nbr));
     for (auto& entry : preload_local_states) {
         local_state.add_application_state(entry.first, entry.second);
```
Does this have strong exception guarantees?

If not, it's better to copy-out local_state, modify it, and copy it back in to avoid an exception from corrupting state.

gleb-cloudius replied:
An exception may occur in `replicate()` call just below. There are two caller to this function: first is during bootstrap and the bootstrap will fail if throws. Second is from the api that enables the gossiper and if the function throws the gossiper stays disabled.

avikivity replied:
I'm talking about add_application_state(). If it fails but partially adjusts local_state, it corrupts state (for example, it will not be replicated).

The existing code uses read-modify-write rather than modify in place to avoid this. I think it's safer to copy.

gleb-cloudius replied:
> I'm talking about add_application_state(). If it fails but partially adjusts local_state, it corrupts state (for example, it will not be replicated).

Yes, and the same will happen if `replicate()` fails.

> 
> The existing code uses read-modify-write rather than modify in place to avoid this. I think it's safer to copy.

It is and if a new version of the series is needed worth adjusting, but I explained why the state does not matter in case the function throws.

avikivity replied:
It's still not good to corrupt state. It's inserting a latent bug.

Relying on global flow to assert that the bug isn't harmful is bad practice. Each component should be correct in its own right and not rely on other components to fail gracefully.

replicate() is careful not to corrupt state, it works in two passes, first allocating updated state, then assigning it.

gleb-cloudius replied:
> It's still not good to corrupt state. It's inserting a latent bug.
> 
> Relying on global flow to assert that the bug isn't harmful is bad practice. Each component should be correct in its own right and not rely on other components to fail gracefully.

The guaranty is that if init succeeded you can use the component if it is not you cannot use it and rely on its internal state, but you can try to init it again. Such interface is common and makes sense.

> 
> replicate() is careful not to corrupt state, it works in two passes, first allocating updated state, then assigning it.

But it will leave the state not replicated if failed. Before the patch the state was either there or not there on all of the shards if replicate() fails, now the state will be on shard0 only. One can argue that not having a local node in gossiper map is also not valid gossiper state, so until gossiper is fully initialized without an error you cannot rely on it state to be valid.

avikivity replied:
It could maybe be excused for init (if the node is shut down on failure), but not for starting gossip from nodetool. I don't understand the insistence on leaving a vulnerable code path where it would be the same effort to make it safe.

gleb-cloudius replied:
> replicate() is careful to be atomic. It will either copy the state to all shards or none.

Except that this patch changes the code to initialize the state in the map on shard 0 and then replicate, so if replication fails shard 0 has the state and others do not.

> 
> Patches should improve the code, not make it worse. Modifying live state makes it worse than it was before compared to read-modify-write.

Even before the patch the code inserted empty state into the map and then if replication failed the sate was broke: zero shard has empty state all other shards have none.

After the patch the worst case is we have partially initialized state on shard0 and none on other shards. How is it worse than before the patch?

The improvement is that shard0 initialize the sate atomically. Which is BTW bags the real issue: the state should not be accessed until fully initialized.

> 
> The fact that we don't reach perfection doesn't mean we should just ignore all aspects of quality.

I do not aware of an aspect of quality that guaranties a consistent object state after initialization failure.

gleb-cloudius replied:
> > > replicate() is careful to be atomic. It will either copy the state to all shards or none.
> > 
> > 
> > Except that this patch changes the code to initialize the state in the map on shard 0 and then replicate, so if replication fails shard 0 has the state and others do not.
> 
> Isn't this wrong? We should maintain atomicity.

Well, yes in case you are going to use gossiper. Does not matter if not.

> 
> > > Patches should improve the code, not make it worse. Modifying live state makes it worse than it was before compared to read-modify-write.
> > 
> > 
> > Even before the patch the code inserted empty state into the map and then if replication failed the sate was broken: zero shard has empty state all other shards have none.
> 
> No, it copied into a local variable, modified it, then replicated atomically.
> 

Did you looks at the code? Look at `my_endpoint_state()`. It emplaces empty state into `_endpoint_state_map` and then returns its copy. The only thing the patch changed it returned a reference to an object in the map.

> > After the patch the worst case is we have partially initialized state on shard0 and none on other shards. How is it worse than before the patch?
> 
> I don't know how to explain why inconsistent state is bad.

For an object that failed initialization and has to be re-initialized before use you cannot explain it because it does not make sense.

> 
> > The improvement is that shard0 initialize the sate atomically. Which is BTW bags the real issue: the state should not be accessed until fully initialized.
> > > The fact that we don't reach perfection doesn't mean we should just ignore all aspects of quality.
> > 
> > 
> > I do not aware of an aspect of quality that guaranties a consistent object state after initialization failure.
> 
> First, it's not just initialization. It's also called when enabling gossip.

This is initialization of the gossiper object. If it fails the gossiper is not enabled and the object has to be re-initialized. I wasn't talking about Scylla initialization.

> 
> Second, a lot of work went into gossip to make it consistent across shards. Violating this because in one path it doesn't matter is wrong. We should not use global knowledge about how gossip is used to relax its invariants.

The patch did not change anything in this regard.

avikivity commented on gms/gossiper.cc:
```diff
@@ -2204,19 +2204,6 @@ future<> gossiper::do_shadow_round(std::unordered_set<gms::inet_address> nodes,
         sleep_abortable(std::chrono::seconds(1), _abort_source).get();
         logger.info("Connect nodes={} again ... ({} seconds passed)",
```
> gossiper: remove duplicate code in do_shadow_round

Ditto here, better to defer cleanups.

calebxyz replied:
Erom Emils commit message "Remove accidentally duplicated code block, likely from a merge conflict." 
it seems that its something that they have introduced by mistake, so i think this one should stay

emdroid replied:
I suppose we can also defer this one - it's a nice one to have, but on the other hand the duplicated code didn't seem to cause issues (the duplication has been there for already quite some time, it wasn't introduced recently) so the commit is not essential for the fix. Thus it should be fine to postpone for later.

emdroid commented on gms/gossiper.cc:
```diff
@@ -2085,6 +2085,15 @@ void gossiper::examine_gossiper(utils::chunked_vector<gossip_digest>& g_digest_l
     }
 }
 
+endpoint_state gossiper::make_initial_endpoint_state(gms::generation_type generation_nbr, const application_state_map& preload_local_states) {
```
Ok, removed this commit, we'll add it separately just for master.

nyh commented on test/cluster/test_gossiper_race.py:
```diff
@@ -0,0 +1,98 @@
+        from_mark=coordinator_log_mark,
+    )
+
+    # test that the coordinator node didn't abort
```
I don't understand what I'm seeing here. Why would the coordinator abort, since you only print a warning and don't abort?

emdroid replied:
Yes, that's true, the comment apparently remained from that version (for 2025.1 that actually did abort) ... now it's checking just for the log message for a repro of the issue (which might or might not abort the node - there is also a check for that at the end of the test that all the servers are still running)

nyh commented on gms/gossiper.cc:
```diff
@@ -1337,6 +1347,12 @@ utils::chunked_vector<gossip_digest> gossiper::make_random_gossip_digest() const
+        // TODO (#25818): re-introduce the on_internal_error() call once all the code paths leading to this are fixed
+        logger.warn("attempting to add a state with empty host id for ip: {}", es.get_ip());
+        co_return;
+    }
```
Is this new diagnostics part of the test (does the test look for it in the log?)? If not, I don't think it belongs in a patch called "add reproducible test".

nyh replied:
The pull request mentions half a dozen issues as being fixed, but not this one...

calebxyz replied:
i think it should mention it as well as it seem to fix exactly the issue it describes, @emdroid am i wrong here?

emdroid replied:
The https://github.com/scylladb/scylladb/issues/25818 might be fixed by the commit 98aaf95ba2efe2d65d0906ecdec7fc8deb6684a4, but if there would still be another code path somewhere that leads to this there would be a danger of possible `abort()` (and breaking the CI again) if we'd reintroduce the `on_internal_error()` here. Therefore it seems to be better to keep this part for this high-priority fix and handle the complete fix for https://github.com/scylladb/scylladb/issues/25818 (if anything's left) later.

emdroid replied:
> Is this new diagnostics part of the test (does the test look for it in the log?)? If not, I don't think it belongs in a patch called "add reproducible test".

Yes, the test specifically looks for this message, as it is used to check whether the issue has been reproduced in the test.

bitpathfinder commented on gms/gossiper.cc:
```diff
@@ -621,6 +621,13 @@ future<> gossiper::do_apply_state_locally(locator::host_id node, endpoint_state
+    // exists in the `_endpoint_state_map` since after a preemption point it could have been deleted.
+    if (!remote_state.get_host_id() && !es) {
+        throw std::runtime_error(format("Entry for host id {} does not exist in the endpoint state map.", node));
+    }
```
Agree, there is a case when is not handled currently: when gossiper performs a shadow round.

bitpathfinder replied:
Added exception handling.

nyh commented on test/cluster/test_gossiper_race.py:
```diff
@@ -15,7 +15,6 @@
 
 @pytest.mark.asyncio
 @skip_mode('release', 'error injections are not supported in release mode')
-@pytest.mark.xfail(reason="https://github.com/scylladb/scylladb/issues/25621")
```
Ok, so I just don't understand why the specific change done in this patch (not the entire PR, just this patch) fixed the xfailing test.

nyh replied:
I'm out of this review. Good luck.

bitpathfinder replied:
Added an explanation why `xfail` was removed.

bitpathfinder commented:
New version: 
* Add exception handling in gossiper `do_shadow_round`, now replies that do not contain a host id will not cause node crashes and all not applied changes in that reply will be skipped.
* Update commit messages descriptions.

emdroid commented:
New version:
* Fixed the comment in the reproducer test about the reproduction scenario ([diff](https://github.com/scylladb/scylladb/compare/1203ca3469d6e5675df2325d70b48fd0be9a1d8a..b34d543f30cb5e80529ed3994f5495157fbec1ae))
